Updating build instructions to include BLAS support

daniandtheweb opened the pull request:
This pull request adds clear instructions about how to build llama.cpp on every platform with and without BLAS support.

daniandtheweb committed:
Updated build information

First update to the build instructions to include BLAS.

daniandtheweb committed:
Update README.md

daniandtheweb committed:
Update information about BLAS

slaren commented:
Looks good, could we also include a brief notice explaining that BLAS is only used when the batch size is at least 32? Making clear that it only benefits prompt processing, but not generation.

For cuBLAS, it may also be useful to point that the CUDA toolkit can be obtained from https://developer.nvidia.com/cuda-downloads

DaniAndTheWeb commented:
Sure.
By installing the CUDA toolkit can Windows users build without any problem? I don't have a Nvidia GPU so I can't test it.

slaren commented:
They also need Visual Studio Community, but yes, they should be able to build with the same cmake command line.

daniandtheweb committed:
Better BLAS explanation

Adding a clearer BLAS explanation and adding a link to download the CUDA toolkit.

slaren commented:
Great! Just one more thing, I suppose it is worth mentioning that on macs, BLAS is already supported by default through the Accelerate framework.

daniandtheweb committed:
Better BLAS explanation

DaniAndTheWeb commented:
Ok, I'll just add that info then

daniandtheweb committed:
BLAS for Mac

Specifying that BLAS is already supported on Macs using the Accelerate Framework.

slaren commented on README.md:
```diff
@@ -184,12 +186,55 @@ Note: For Windows, CMake or Zig can be used.
 
+#### BLAS Build
+
+Building the program with BLAS support may lead to some performance improvements in prompt processing using batch sizes higher than 32 (the default is 512). BLAS doesn't affect the prompt generation. There are currently three different implementations of it:
```
"BLAS doesn't affect the prompt generation" seems wrong, something like "BLAS doesn't affect the normal generation performance" may be clearer.

daniandtheweb committed:
Clarify the effect of BLAS

slaren commented:
The batch still has to be at least 32 tokens, it won't be used with smaller prompts.

SlyEcho commented:
BTW, @DaniAndTheWeb yesterday I gave some messy instructions on how to get OpenBLAS llama.cpp on Windows: https://github.com/ggerganov/llama.cpp/discussions/1153

DaniAndTheWeb commented:
@SlyEcho Producing the build instructions for Windows in a clear way seems to add quite a lot of space to the readme. If it's not a problem I can just add all the needed instructions.

DaniAndTheWeb commented:
@SlyEcho I'm finishing a revised version of the build instructions to include make on Windows. Is there any reason why you recommended the fortran version of w64devkit?

SlyEcho commented:
It doesn't really hurt anything. The reason I recommend w64devkit, is that it is just download and extract and it's ready to use. No need to install anything, nothing to clean up, just delete when you don't need it.

daniandtheweb committed:
Windows Make instructions

Added the instructions to build with Make on Windows

daniandtheweb committed:
Fixing typo

DaniAndTheWeb commented:
I think that it may be quite clear like this.

slaren commented on README.md:
````diff
@@ -184,12 +196,71 @@ Note: For Windows, CMake or Zig can be used.
+          make LLAMA_OPENBLAS=1
+          ```
+
+  - Using `CMake` on Linux:
````
cmake should work everywhere, I don't think that we need to specify "on Linux" here

DaniAndTheWeb replied:
I've specified it because I'm unable to find a guide that includes OpenBLAS on Windows using CMake. If there's a way to do it I'd be happy to include it in the README.

slaren replied:
I gave it a try by installing OpenBLAS with vcpkg, but it doesn't work, the CMakeLists.txt would need to be updated to support building with OpenBLAS on Windows. So I think this is fine for now.

slaren replied:
After installing OpenBLAS with vcpkg, it shows this:
```
openblas provides CMake targets:

    # this is heuristically generated, and may not be correct
    find_package(OpenBLAS CONFIG REQUIRED)
    target_link_libraries(main PRIVATE OpenBLAS::OpenBLAS)
```
So maybe that would work better.

SlyEcho replied:
```cmake
if (LLAMA_OPENBLAS)
    
	find_package(OpenBLAS)
	
    if (OpenBLAS_FOUND)
        message(STATUS "OpenBLAS found")

        add_compile_definitions(GGML_USE_OPENBLAS)
        set(LLAMA_EXTRA_LIBS ${LLAMA_EXTRA_LIBS} ${OpenBLAS_LIBRARIES})

        include_directories(${OpenBLAS_INCLUDE_DIRS})
    else()
        message(WARNING "OpenBLAS not found")
    endif()
endif()
```
This works with the OpenBLAS release .zip, if you pass the path to that on `CMAKE_PREFIX_PATH` but the CMake files that it comes with need to be edited to fix the paths inside.

SlyEcho replied:
In Linux then it doesn't work because cblas is a separate library :disappointed_relieved:

slaren replied:
This works on Windows with vcpkg and seems clean enough:
```cmake
    find_package(OpenBLAS)
    if (OpenBLAS_FOUND)
        message(STATUS "OpenBLAS found")
        add_compile_definitions(GGML_USE_OPENBLAS)
        set(LLAMA_EXTRA_LIBS ${LLAMA_EXTRA_LIBS} OpenBLAS::OpenBLAS)
    else()
        message(WARNING "OpenBLAS not found")
    endif()
```
But for whatever reason, the Ubuntu package `libopenblas-dev` doesn't include the cmake module, even though it has been on the OpenBLAS repository [for 5 years](https://github.com/xianyi/OpenBLAS/blob/develop/cmake/OpenBLASConfig.cmake.in).

I give up, the performance difference isn't worth it anyway, and eventually our mat mul routines should be faster than OpenBLAS.

SlyEcho replied:
There is the hard way, we make a `cmake/` subdirectory and add that to `CMAKE_MODULE_PATH` and inside we keep a bunch of scripts like `FindOpenBLAS.cmake` that just try all kinds of places to find the libraries, headers etc.

daniandtheweb committed:
Fix trailing whitespace
